packages/definitions-parser/src/git.ts

```diff
@@ -66,62 +56,74 @@ export async function gitDiff(log: Logger, definitelyTypedPath: string): Promise
     return stdout;
   }
 }
-
-/** Returns all immediate subdirectories of the root directory that have changed. */
-export function gitChanges(diffs: GitDiff[]): PackageId[] {
-  const changedPackages = new Map<string, Map<string, DependencyVersion>>();
-
+/** Returns all immediate subdirectories of the root directory that have been deleted. */
+function gitDeletions(diffs: GitDiff[]): { errors: string[] } | { ok: PackageId[] } {
+  const changedPackages = new Map<string, PackageId>();
+  const errors = [];
   for (const diff of diffs) {
+    if (diff.status !== "D") continue;
     const dep = getDependencyFromFile(diff.file);
     if (dep) {
-      const versions = changedPackages.get(dep.name);
-      if (!versions) {
-        changedPackages.set(dep.name, new Map([[formatDependencyVersion(dep.version), dep.version]]));
-      } else {
-        versions.set(formatDependencyVersion(dep.version), dep.version);
-      }
+      const key = `${dep.typesDirectoryName}/v${formatDependencyVersion(dep.version)}`;
+      changedPackages.set(key, dep);
+    } else {
+      errors.push(
+        `Unexpected file deleted: ${diff.file}
+When removing packages, you should only delete files that are a part of removed packages.`
+      );
     }
   }
-
-  return Array.from(
-    flatMapIterable(changedPackages, ([name, versions]) => mapIterable(versions, ([_, version]) => ({ name, version })))
-  );
+  return errors.length ? { errors } : { ok: Array.from(changedPackages.values()) };
 }
 
+function formatDependencyVersion(version: DirectoryParsedTypingVersion | "*") {
+  return version === "*" ? "*" : formatTypingVersion(version);
+}
 export async function getAffectedPackagesFromDiff(
-  dt: FS,
+  allPackages: AllPackages,
   definitelyTypedPath: string,
   selection: "all" | "affected" | RegExp
-) {
-  const allPackages = await AllPackages.read(dt);
+): Promise<string[] | PreparePackagesResult> {
+  const errors = [];
   const diffs = await gitDiff(consoleLogger.info, definitelyTypedPath);
   if (diffs.find((d) => d.file === "notNeededPackages.json")) {
-    for (const deleted of getNotNeededPackages(allPackages, diffs)) {
-      checkNotNeededPackage(deleted);
+    const deleteds = getNotNeededPackages(allPackages, diffs);
+    if ("errors" in deleteds) errors.push(...deleteds.errors);
+    else
+      for (const deleted of deleteds.ok as NotNeededPackage[]) {
+        errors.push(...(await checkNotNeededPackage(deleted)));
```

Comment:
Not proposing to change it now, but it’s really weird that this `get*` function is _checking_ something as a side effect. It’s also weird that a function that’s supposed to “get affected packages” takes a parameter for whether you want... the affected packages... or all of them 🤔

Reply:
I also just noticed that `selection` can be a regex, and that there's code to run only packages that match it.

Reply:
Moments later: This code is **only** called with "affected"; "all" and regexes are unused. I deleted the dead code.